From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shengjiu Wang <shengjiu.wang@nxp.com>
Date: Wed, 18 Mar 2026 18:28:49 +0800
Subject: ASoC: simple-card-utils: Check value of is_playback_only and
 is_capture_only

Backport of upstream 0e9fc79132ce. simple-card.c parses the link direction
on the card root, then again on the parent, CPU and codec nodes. The CPU and
codec nodes do not carry playback-only, so they clear the root-level flag and
the sun55i-a523 HDMI card stays capture-capable. This is already fixed on 7.2
by the same upstream change; 6.18.y stable does not have it.

Signed-off-by: Shengjiu Wang <shengjiu.wang@nxp.com>
---
 sound/soc/generic/simple-card-utils.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/sound/soc/generic/simple-card-utils.c b/sound/soc/generic/simple-card-utils.c
index 111111111111..222222222222 100644
--- a/sound/soc/generic/simple-card-utils.c
+++ b/sound/soc/generic/simple-card-utils.c
@@ -1223,9 +1223,9 @@ void graph_util_parse_link_direction(struct device_node *np,
 	bool is_playback_only = of_property_read_bool(np, "playback-only");
 	bool is_capture_only  = of_property_read_bool(np, "capture-only");
 
-	if (np && playback_only)
+	if (playback_only && is_playback_only)
 		*playback_only = is_playback_only;
-	if (np && capture_only)
+	if (capture_only && is_capture_only)
 		*capture_only = is_capture_only;
 }
 EXPORT_SYMBOL_GPL(graph_util_parse_link_direction);
-- 
Armbian

